plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesSplitSource.java

```diff
@@ -62,71 +62,107 @@ public TableChangesSplitSource(
+                .filter(Objects::nonNull)
+                .findFirst()
+                .orElseThrow();
     }
 
     @Override
     public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize) // TODO dont ignore maxSize https://github.com/trinodb/trino/issues/17182
     {
         long processedVersion = currentVersion.getAndIncrement();
-        try {
-            if (processedVersion > tableReadVersion) {
-                return CompletableFuture.completedFuture(new ConnectorSplitBatch(ImmutableList.of(), true));
-            }
-            List<DeltaLakeTransactionLogEntry> entries = getEntriesFromJson(processedVersion, transactionLogDir, fileSystem)
-                    .orElseThrow(() -> new TrinoException(DELTA_LAKE_BAD_DATA, "Delta Lake log entries are missing for version " + processedVersion));
-            if (entries.isEmpty()) {
-                return CompletableFuture.completedFuture(new ConnectorSplitBatch(ImmutableList.of(), false));
+        if (processedVersion > tableReadVersion) {
+            return CompletableFuture.completedFuture(new ConnectorSplitBatch(ImmutableList.of(), true));
+        }
+        List<DeltaLakeTransactionLogEntry> entries = getEntries(processedVersion, transactionLogDir, fileSystem);
+        if (entries.isEmpty()) {
+            return CompletableFuture.completedFuture(new ConnectorSplitBatch(ImmutableList.of(), false));
+        }
+        CommitInfoEntry commitInfo = extractCommitInfoEntry(entries);
+        List<ConnectorSplit> splits = new ArrayList<>();
+        boolean containsCdcEntry = false;
+        boolean containsRemoveEntry = false;
+        for (DeltaLakeTransactionLogEntry entry : entries) {
+            CdcEntry cdcEntry = entry.getCDC();
+            if (cdcEntry != null) {
+                containsCdcEntry = true;
+                splits.add(mapToDeltaLakeTableChangesSplit(
+                        commitInfo,
+                        CDF_FILE,
+                        cdcEntry.getSize(),
+                        cdcEntry.getPath(),
+                        cdcEntry.getCanonicalPartitionValues()));
             }
-            List<CommitInfoEntry> commitInfoEntries = entries.stream()
-                    .map(DeltaLakeTransactionLogEntry::getCommitInfo)
-                    .filter(Objects::nonNull)
-                    .collect(toImmutableList());
-            if (commitInfoEntries.size() != 1) {
-                throw new TrinoException(DELTA_LAKE_BAD_DATA, "There should be exactly 1 commitInfo present in a metadata file");
+            if (entry.getRemove() != null && entry.getRemove().isDataChange()) {
+                containsRemoveEntry = true;
             }
-            CommitInfoEntry commitInfo = getOnlyElement(commitInfoEntries);
-            List<ConnectorSplit> splits = new ArrayList<>();
-            boolean containsCdcEntry = false;
-            boolean containsRemoveEntry = false;
+        }
+        if (containsRemoveEntry && !containsCdcEntry) {
```

Comment:
Extract refactors to a prep commit. Otherwise the diff consist of new code being added and many lines moving around. It is hard for me to understand what are the important changes I should focus my review on.

Reply:
It is not a refactor per se, some code was moved to constructor because  now currentVersion is calculated based on firstReadVersion or firstReadTimestamp

Reply:
Is it possible to still separate refactor-like changes (not changing semantics) from others?